File: LeetCode/twopoint/StringCompression.java
package twopoint;

public class StringCompression {
    public int compress(char[] chars) {
        int n = chars.length;
        int read = 0, write = 0;

        while (read < n) {
            char ch = chars[read];
            int count = 0;

            while (read < n && chars[read] == ch) {
                count++;
                read++;
            }

            chars[write++] = ch;

            if (count > 1) {
                for (char c : Integer.toString(count).toCharArray()) {
                    chars[write++] = c;
                }
            }
        }

        return write;
    }

    public static void main(String[] args) {
        StringCompression solution = new StringCompression();
        char[] chars1 = new char[]{'a', 'a', 'b', 'b', 'c', 'c', 'c'};
        System.out.println(solution.compress(chars1)); // Output: 6

        char[] chars2 = new char[]{'a'};
        System.out.println(solution.compress(chars2)); // Output: 1

        char[] chars3 = new char[]{'a', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b'};
        System.out.println(solution.compress(chars3)); // Output: 4
    }
}

File: LeetCode_Python/twopoint/StringCompression.py
class StringCompression:
    def compress(self, chars: list[str]) -> int:
        n = len(chars)
        read = write = 0

        while read < n:
            char = chars[read]
            start = read

            while read < n and chars[read] == char:
                read += 1

            chars[write] = char
            write += 1

            count = read - start
            if count > 1:
                for digit in str(count):
                    chars[write] = digit
                    write += 1

        return write


# Time Complexity: O(n)
# Space Complexity: O(1)

if __name__ == "__main__":
    solution = StringCompression()

    print("Expected Output: 6")
    print("Actual Output: ", solution.compress(["a", "a", "b", "b", "c", "c", "c"]))

    print("Expected Output: 4")
    print(
        "Actual Output: ",
        solution.compress(
            ["a", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b", "b"]
        ),
    )

    print("Expected Output: 1")
    print("Actual Output: ", solution.compress(["a"]))
